File: packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/radix-tabs-trigger-requires-list.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { findRequiredAncestorPlacement } from "../../utils/find-required-ancestor-placement.js";
import { resolveNamespacedPartName } from "../../utils/resolve-namespaced-part-name.js";
import { resolveShadcnUiComponentName } from "../../utils/resolve-shadcn-ui-component-name.js";
import type { RuleContext } from "../../utils/rule-context.js";

const RADIX_TABS_PRIMITIVE_MODULE_PATTERN = /^@radix-ui\/react-tabs$/;
const RADIX_UNIFIED_MODULE_PATTERN = /^radix-ui$/;

const resolveRadixTabsPartName = (elementName: EsTreeNode, context: RuleContext): string | null =>
  resolveShadcnUiComponentName(elementName, RADIX_TABS_PRIMITIVE_MODULE_PATTERN, context) ??
  resolveNamespacedPartName(elementName, RADIX_UNIFIED_MODULE_PATTERN, "Tabs", context);

export const radixTabsTriggerRequiresList = defineRule({
  id: "radix-tabs-trigger-requires-list",
  title: "Radix tabs trigger is outside Tabs.List",
  severity: "warn",
  category: "Correctness",
  requires: ["radix-ui"],
  matchByOccurrence: true,
  recommendation:
    "Render each Radix Tabs.Trigger inside Tabs.List so the tablist role and roving keyboard focus cover every trigger.",
  create: (context: RuleContext) => ({
    JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
      if (resolveRadixTabsPartName(node.name, context) !== "Trigger") return;
      const placement = findRequiredAncestorPlacement(node, (ancestorName) => {
        const resolvedName = resolveRadixTabsPartName(ancestorName, context);
        if (resolvedName === "List") return "required";
        return resolvedName === "Root" ? "root" : null;
      });
      if (placement !== "inside-root-without-required") return;
      context.report({
        node,
        message:
          "This Tabs.Trigger is outside Tabs.List, so it misses the tablist grouping and roving keyboard focus Radix provides through the list. Nest it inside Tabs.List.",
      });
    },
  }),
});

File: native/oxlint/rules/radix-tabs-trigger-requires-list.rs
use crate::{context::LintContext, rule::Rule, AstNode};
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

const MESSAGE: &str = "This Tabs.Trigger is outside Tabs.List, so it misses the tablist grouping and roving keyboard focus Radix provides through the list. Nest it inside Tabs.List.";

#[derive(Debug, Default, Clone)]
pub struct RadixTabsTriggerRequiresList;

declare_oxc_lint!(
    /// Require Radix tabs triggers to be inside Tabs.List.
    RadixTabsTriggerRequiresList,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Require Radix tabs triggers to be inside Tabs.List.",
);

impl Rule for RadixTabsTriggerRequiresList {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        if !has_capability_or_unspecified(ctx, "radix-ui") {
            return;
        }
        let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
            return;
        };
        if resolve_radix_tabs_part_name(&opening_element.name, ctx) != Some("Trigger")
            || !jsx_part_is_inside_root_without_required_ancestor(node, ctx, |ancestor_name| {
                match resolve_radix_tabs_part_name(ancestor_name, ctx) {
                    Some("List") => Some(JsxPartAncestorClassification::Required),
                    Some("Root") => Some(JsxPartAncestorClassification::Root),
                    _ => None,
                }
            })
        {
            return;
        }
        ctx.diagnostic(OxcDiagnostic::warn(MESSAGE).with_label(opening_element.span));
    }
}

fn resolve_radix_tabs_part_name<'a>(
    element_name: &oxc_ast::ast::JSXElementName<'a>,
    ctx: &LintContext<'a>,
) -> Option<&'static str> {
    if let Some(api_path) = resolve_jsx_import_api_path(
        element_name,
        |module_source| module_source == "@radix-ui/react-tabs",
        ctx,
    ) {
        let [part_name] = api_path.as_slice() else {
            return None;
        };
        return radix_tabs_part(part_name);
    }
    let api_path = resolve_jsx_import_api_path(
        element_name,
        |module_source| module_source == "radix-ui",
        ctx,
    )?;
    let [namespace, part_name] = api_path.as_slice() else {
        return None;
    };
    if namespace != "Tabs" {
        return None;
    }
    radix_tabs_part(part_name)
}

fn radix_tabs_part(part_name: &str) -> Option<&'static str> {
    match part_name {
        "Trigger" => Some("Trigger"),
        "List" => Some("List"),
        "Root" => Some("Root"),
        _ => None,
    }
}
